cbmc/proofs/mqtt_state.c

```diff
@@ -0,0 +1,560 @@
+ * IotMqttSubscription array list
+ ****************************************************************/
+
+IotMqttSubscription_t *allocate_IotMqttSubscriptionArray( IotMqttSubscription_t *pSub,
+							  size_t length )
+{
+  if ( pSub == NULL ) pSub = malloc_can_fail( length * sizeof( *pSub ) );
+  if ( pSub == NULL ) return NULL;
+
+  for ( int i = 0; i < length; i++ )
+    allocate_IotMqttSubscription( pSub + i );
+
+  return pSub;
+}
+
+bool valid_IotMqttSubscriptionArray( const IotMqttSubscription_t *pSub,
+				     const size_t length )
+{
+  if ( !IFF( pSub == NULL, length == 0 ) ) return false;
+  if ( pSub == NULL ) return false;
+
+  bool result = 1;
+  for ( int i = 0; i < length; i++ )
+    result = result && valid_IotMqttSubscription( pSub + i );
+
+  return
+#ifdef SUBSCRIPTION_COUNT_MAX
+    length < SUBSCRIPTION_COUNT_MAX &&
+#endif
+    result;
+}
+
+/****************************************************************
+ * IotMqttSubscription array list
+ ****************************************************************/
+
+// Subscription linked list elements
+
+_mqttSubscription_t *allocate_IotMqttSubscriptionListElt( _mqttSubscription_t *pElt )
+{
+  size_t length;
+  // Assumption avoids arithmetic overflow in malloc, rechecked in valid_*
+  __CPROVER_assume( length < CBMC_MAX_OBJECT_SIZE - sizeof( *pElt ) );
+
+  if ( pElt == NULL ) pElt = malloc_can_fail( sizeof( *pElt ) + length );
+  if ( pElt == NULL ) return NULL;
+
+  pElt->link.pPrevious = NULL;
+  pElt->link.pNext = NULL;
+  pElt->topicFilterLength = length;
+  return pElt;
+}
+
+bool valid_IotMqttSubscriptionListElt( const _mqttSubscription_t *pElt )
+{
+  if ( pElt == NULL ) return false;
+
+  return
+#ifdef TOPIC_LENGTH_MAX
+    pElt->topicFilterLength < TOPIC_LENGTH_MAX &&
+#endif
+    pElt->topicFilterLength < CBMC_MAX_OBJECT_SIZE - sizeof( *pElt ) &&
+    __CPROVER_r_ok( pElt->pTopicFilter, pElt->topicFilterLength ) &&
+    __CPROVER_w_ok( pElt->pTopicFilter, pElt->topicFilterLength );
+}
+
+// Subscription linked lists
+
+IotListDouble_t *allocate_IotMqttSubscriptionList( IotListDouble_t *pSub,
+						   size_t length )
+{
+  if ( pSub == NULL ) pSub = malloc_can_fail( sizeof( *pSub ) );
+  if ( pSub == NULL ) return NULL;
+
+  IotListDouble_Create( pSub );
+  for ( int i = 0; i < length; i++ ) {
+    _mqttSubscription_t *pElt = allocate_IotMqttSubscriptionListElt( NULL );
+    __CPROVER_assume( pElt );
+    IotListDouble_InsertHead( pSub, &( pElt->link ) );
+  }
+  return pSub;
+}
+
+bool valid_IotMqttSubscriptionList( const IotListDouble_t *pSub,
+				    const size_t length )
+{
+  if ( pSub == NULL ) return false;
+
+  bool result = 1;
+  IotListDouble_t *pLink;
+  IotContainers_ForEach( pSub, pLink ) {
+    _mqttSubscription_t
+      *pElt = IotLink_Container( _mqttSubscription_t, pLink, link );
+    result = result && valid_IotMqttSubscriptionListElt( pElt );
+  }
+
+  return
+#ifdef SUBSCRIPTION_COUNT_MAX
+    length < SUBSCRIPTION_COUNT_MAX &&
```

Comment:
`<` or `<=`<?

Reply:
<.  MAX is the value used in the loop unwinding and is one greater than the maximum length.  I've added comments everywhere.  We could choose better names on a later revision.